pipeline/pipeline.proto

```diff
@@ -47,91 +64,107 @@ service Pipeline {
-
   string page_token = 2 [(google.api.field_behavior) = OPTIONAL];
+  View view = 3 [(google.api.field_behavior) = OPTIONAL];
 
-  bool with_recipe = 3 [(google.api.field_behavior) = OPTIONAL];
-
-  bool query = 4 [(google.api.field_behavior) = OPTIONAL];
+  enum View {
+    VIEW_UNSPECIFIED = 0;
+    BASIC = 1;
+    WITH_RECIPE = 2;
+  }
 }
 
 message ListPipelinesResponse {
-  repeated PipelineInfo pipelines = 1;
+  repeated PipelineInfo contents = 1;
   string next_page_token = 2;
 }
 
 message GetPipelineRequest {
-  string name = 1;
+  string name = 1 [(google.api.field_behavior) = REQUIRED];
 }
 
 message UpdatePipelineRequest {
-  string name = 1 [(google.api.field_behavior) = REQUIRED];
-  string description = 2 [(google.api.field_behavior) = OPTIONAL];
-  bool active = 3 [(google.api.field_behavior) = OPTIONAL];
-  instill.pipeline.Recipe recipe = 4 [(google.api.field_behavior) = OPTIONAL];
+  PipelineInfo pipeline = 1 [(google.api.field_behavior) = REQUIRED];
 
-  google.protobuf.FieldMask update_mask = 5 [(google.api.field_behavior) = REQUIRED];
+  google.protobuf.FieldMask update_mask = 2 [(google.api.field_behavior) = REQUIRED];
 }
 
 message DeletePipelineRequest {
-  string name = 1;
+  string name = 1 [(google.api.field_behavior) = REQUIRED];
 }
 
 message TriggerPipelineContent {
   string url = 1 [(google.api.field_behavior) = OPTIONAL];
   string base64 = 2 [(google.api.field_behavior) = OPTIONAL];
+  bytes chunk = 3 [(google.api.field_behavior) = OPTIONAL];
 }
 
 message TriggerPipelineRequest {
-  string name = 1;
-  repeated TriggerPipelineContent contents = 2;
+  string name = 1 [(google.api.field_behavior) = REQUIRED];
+  repeated TriggerPipelineContent contents = 2 [(google.api.field_behavior) = REQUIRED];
+}
+
+message Scheduler {
+  string crontab = 1 [(google.api.field_behavior) = REQUIRED];
 }
 
-message DataSource {
-  string type = 1;
+message Source {
+  string type = 1 [(google.api.field_behavior) = REQUIRED];
+  string name = 2 [(google.api.field_behavior) = REQUIRED];
+  Scheduler scheduler = 3 [(google.api.field_behavior) = OPTIONAL];
 }
 
-message DataDestination {
-  string type = 1;
+message Destination {
+  string type = 1 [(google.api.field_behavior) = REQUIRED];
+  string name = 2 [(google.api.field_behavior) = REQUIRED];
 }
 
-message VisualDataOperator {
-  string model_id = 1;
-  int32 version = 2;
+message Model {
+  string name = 1 [(google.api.field_behavior) = REQUIRED];
+  int32 version = 2 [(google.api.field_behavior) = REQUIRED];
 }
 
 message LogicOperator {
 }
 
 message Recipe {
-  DataSource data_source = 1;
-  DataDestination data_destination = 2;
-  repeated VisualDataOperator visual_data_operator = 3;
-  repeated LogicOperator logic_operator = 4;
+  Source source = 1 [(google.api.field_behavior) = REQUIRED];
+  Destination destination = 2 [(google.api.field_behavior) = REQUIRED];
+  repeated Model model = 3 [(google.api.field_behavior) = REQUIRED];
```

Comment:
@bochengyang @pinglin recipe will create a sequence of models. Do we know how the output of 1st model fits with the input of the next model?

Reply:
currently, pipeline-backend only fetch the first array of models, I’m not sure if we might support multi-model’s operation, so I leave the flexibility here once we need a multi-model operation

Reply:
We support only a single model (i.e., an ensemble model in Triton), single source and single destination for now. 

We will decide whether to support multiple models in parallel or subsequent models from a single source later.